### PR TITLE
Apply some post-processing over the multi ADB-name PR

### PR DESCRIPTION
**Pull request by d4vidi**
- [ ] This is a small change 
- [ ] This change has been discussed in issue #<?> and the solution has been agreed upon with maintainers.

---

**Description:**

This is a direct continuation of #2169. Mostly, but not solely, focused on improving unit tests and coverage rate.

### CODE REVIEW

File: detox/src/devices/drivers/android/exec/ADB.js [hunk 1]
```diff
@@ -297,22 +289,4 @@ class ADB {
...
-    `------ BEGIN DUMP ------`,
-    `adb devices base64: ${encodeBase64(stdout)}`,
-    `adb devices: ${stdout}`,
-    `port: ${device.port}`,
-    `------  END DUMP  ------`,
-  ].join('\n');
-
-  throw new Error(errorMessage);
-}
-
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/exec/ADB.js:317]: @rotemmiz @noomorph Friends, I think it's been long enough.

File: detox/src/devices/drivers/android/tools/EmulatorHandle.js [hunk 2]
```diff
@@ -5,25 +5,23 @@ class EmulatorHandle extends DeviceHandle {
...
 
+    this.telnet = new EmulatorTelnet();
     this.port = this.adbName.split('-')[1];
   }
 
-  queryName() {
+  async queryName() {
     if (!this._name) {
-      this._name = this._queryNameViaTelnet();
+      this._name = await this._queryNameViaTelnet();
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/EmulatorHandle.js:14]: So was this a bug, or not? (not related to #2169).

File: detox/src/devices/drivers/android/tools/FreeDeviceFinder.js [hunk 3]
```diff
@@ -11,28 +11,38 @@ class FreeDeviceFinder {
   async findFreeDevice(deviceQuery) {
     const { devices } = await this.adb.devices();
     for (const candidate of devices) {
-      const isBusy = this.deviceRegistry.isDeviceBusy(candidate.adbName);
-      if (isBusy) {
-        log.debug({ event: DEVICE_LOOKUP_LOG_EVT }, `Device ${candidate.adbName} is busy, skipping...`);
-        continue;
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/FreeDeviceFinder.js:17]: Sorry, had to change this as I'm not a fan of `continue`

File: detox/src/devices/drivers/android/tools/__mocks__/handles.js [hunk 4]
```diff
@@ -1,22 +1,21 @@
...
-class MockEmulatorHandle extends EmulatorHandle {
-  queryName() {
-    return Promise.resolve(mockAvdName);
-  }
-}
 
-const emulator5556 = new MockEmulatorHandle('emulator-5556\tdevice');
-const localhost5555 = new DeviceHandle('localhost:5555\tdevice');
-const ip5557 = new DeviceHandle('192.168.4.19:5557\tdevice');
-const deviceOffline = new MockEmulatorHandle('192.168.4.19:5559\toffline');
```
Comment (d4vidi) [inline detox/src/devices/drivers/android/tools/__mocks__/handles.js:14]: I believe in pure mocking - avoiding the coupling of this scope with the logic of Handlers (namely, the format of the accepted string).

File: detox/src/devices/drivers/android/tools/EmulatorHandle.js (hunk 2 above)
Comment (noomorph) [reply detox/src/devices/drivers/android/tools/EmulatorHandle.js:14]: @d4vidi , no, don't do it!!! 😱

Comment (noomorph) [reply detox/src/devices/drivers/android/tools/EmulatorHandle.js:14]: I am caching a promise inside `this._name` and making sure to return always the same instance of a Promise. Please don't touch this code.

File: detox/src/devices/drivers/android/tools/FreeDeviceFinder.js [hunk 5]
```diff
@@ -11,28 +11,38 @@ class FreeDeviceFinder {
...
+      if (await this._isDeviceFreeAndMatching(candidate, deviceQuery)) {
+        log.debug({ event: DEVICE_LOOKUP_LOG_EVT }, `Found a matching & free device ${candidate.adbName}`);
```
Comment (noomorph) [inline detox/src/devices/drivers/android/tools/FreeDeviceFinder.js:15]: Can't we have all those `log.debug` calls inside `_isDeviceFreeAndMatching` function?

File: detox/src/devices/drivers/android/tools/__mocks__/handles.js (hunk 4 above)
Comment (noomorph) [reply detox/src/devices/drivers/android/tools/__mocks__/handles.js:14]: 🤦‍♂️ okay, whatever

File: detox/src/devices/drivers/android/exec/ADB.js (hunk 1 above)
Comment (noomorph) [reply detox/src/devices/drivers/android/exec/ADB.js:317]: Agree.

File: detox/src/devices/drivers/android/tools/FreeDeviceFinder.js (hunk 5 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/tools/FreeDeviceFinder.js:15]: Yes, we can!

File: detox/src/devices/drivers/android/tools/__mocks__/handles.js (hunk 4 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/tools/__mocks__/handles.js:14]: But that's a problem in both approaches

Comment (d4vidi) [reply detox/src/devices/drivers/android/tools/__mocks__/handles.js:14]: Actually, it is not as sever. I'll reconsider this change.

File: detox/src/devices/drivers/android/tools/EmulatorHandle.js (hunk 2 above)
Comment (d4vidi) [reply detox/src/devices/drivers/android/tools/EmulatorHandle.js:14]: np, I'll revert - but with the subtle change of renaming `_name` to `_namePromise`.
Is it really that important to cache `name` as a promise and not as the resulted value?

**Pull request merged**